Implement DTLSSocket and fix non-blocking connections on TLSSocket

SeppoTakalo opened the pull request:
### Description

Implement DTLSSocketWrapper and fix non-blocking connections on TLSSocket

DTLSSocketWrapper is equivalent of TLSSocketWrapper but uses datagram mode
and timers for handling Mbed TLS timeouts.
    
Non-blocking connections were not working earlier, now fixed for both
secure socket modes.

This requires changes once https://github.com/ARMmbed/mbed-os/pull/8651 goes in, so pending for that one, or if we merge this before, I need to rework the #8651 

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

Commit:
Implement DTLSSocketWrapper and fix non-blocking connections on TLSSocket

DTLSSocketWrapper is equivalent of TLSSocketWrapper but uses datagram mode
and timers for handling Mbed TLS timeouts.

Non-blocking connections were not working earlier, now fixed for both
secure socket modes.

Commit:
Implement DTLSSocket helper

This is equivalent of TLSSocket class but for UDP and DTLS.

SeppoTakalo commented:
@kjbracey-arm Please review

michalpasztamobica committed:
Add unit tests for TLSSocket and TLSSocketWrapper

This also includes their Datagram counterparts: DTLSSocket and DTLSSocketWrapper.
Coverage missing for timer-related functionality, but this would require more advanced stub development.
The extra (d)tls_test_config.h is necessary for successful compilation without HW support for the mbedtls features in unittests.
I reused the mbedtls stub found in features/nanostack/coap-service/test/coap-service/unittest/stub/mbedtls_stub.c and amended it slightly with a few missing functions.

cmonr commented:
@kjbracey-arm When you get the chance, this needs a review.
@SeppoTakalo When you get a chance, this'll need a rebase.

cmonr commented:
@SeppoTakalo We'll be aiming to take in #8651 first since its last test failures were not due to the PR.

kjbracey commented on features/netsocket/DTLSSocket.cpp:
```diff
@@ -0,0 +1,51 @@
+    close();
+}
+
+#endif // MBEDTLS_SSL_CLI_C
```
Missing newline. Same in lots of other files

kjbracey commented on features/netsocket/DTLSSocketWrapper.cpp:
```diff
@@ -0,0 +1,55 @@
+
+    if (context->_timer_event_id == 0) {
+        return -1;
+    } else if (context->_timer_expired) {
```
Seems current implementation of mbed TLS doesn't rely on the intermediate delay, but nothing in the documentation says you can skip it. I'd suggest putting it in for completeness in a later patch.

kjbracey reviewed (approved):
Seems fine, just formatting nits and note for the future.

Commit:
Decrease verbosity of the TLSSocketWrapper

Commit:
Refactor DTLSSocket to use Socket::getpeername()

Also, let DTLSSocket to be a friend of InternetSocket so it can do
the name resolution from its _stack.

+ Some whitespace fixes

Commit:
Need to protect members with MBEDTLS_PK_C and X509_CRT_PARSE flags.

This is dependency to X509_CRT_PARSE but just to make sure
that also the 'mbedtls_pk_context' structure is not used if that
flag is not enabled.

Commit:
Compare the intermediate timestamps as well

SeppoTakalo commented:
* Unittests added and fixed.
* whitespace/line endings fixed
* rebased on top of master
* Added comparison for intermediate timeouts
* Manually tested against DTLS server from Mbed TLS
* unittests passing.
* Refactored the DTLSSocket to use `UDPSocket::getpeername()` instead of having copy of the endpoint address.


```
+---------------+---------+-------------+-------------+-----------+-----------+
| Testcase      | Verdict | Fail Reason | Skip Reason | platforms |  duration |
+---------------+---------+-------------+-------------+-----------+-----------+
| DTLS          |   pass  |             |             |    K64F   | 11.730098 |
| DTLS_NONBLOCK |   pass  |             |             |    K64F   | 12.998959 |
+---------------+---------+-------------+-------------+-----------+-----------+
```

cmonr commented:
/morph build

cmonr commented:
Export falures do not appear to be due to PR.
(Jenkins Nodes dropped)

Will restart when able.

SeppoTakalo commented:
Unrelated test failure:
`tests-mbed_hal-rtc_time_conv`

cmonr commented:
@SeppoTakalo Yup. Will restart when able.

0xc0170 commented:
This failure is tracked in the issues and one potentional fix was created as I understood. We will restart CI as soon as possible (one rollup now in there). We might create another rollup of few 5.11 unrelated PR (this would be a good candidate to be included).

cmonr commented:
@SeppoTakalo Fyi, in trying to create https://github.com/ARMmbed/mbed-os/pull/8760 (Rollup PR) with this PR (to expidite retesting while keeping CI load low), I found that a merge conflict is created when this PR is merged with https://github.com/ARMmbed/mbed-os/pull/8647. 

I've put 8647 in instead of this PR because 1) the other PR is a day older, and 2) I suspect that rebase resolution will be more straightforward with this PR than for 8647.

_TL;DR: Once #8760 passes and is merged, this PR will need a rebase because of a merge conflict with #8647._

Commit:
Clarify that TLSSocket should not be reopened.

SeppoTakalo commented:
Rebased

0xc0170 commented:
Can you look at styling change failures (don't mind .cmake, we identified the bug in the grep and fix is already in the CI)

0xc0170 commented:
@SeppoTakalo If you rebase, master contains the fix that addresses travis failure and this will be ready for CI

Commit:
Astyle fixes

SeppoTakalo commented:
Rebased and fixed the Astyle findings.

Commit:
Include TLSSocket and DTLSSocket in Doxygen

0xc0170 commented on UNITTESTS/features/netsocket/DTLSSocket/dtls_test_config.h:
```diff
@@ -0,0 +1,11 @@
+/*
```
Some configs in this PR do not contain the license, please add

0xc0170 commented:
Looks good , while I was preparing another rollup PR I noticed some license headers are missing, please add them.

Update: below close was just unfortunate misclick, sorry about that

SeppoTakalo commented:
will add.. just a sec

Commit:
Add license headers into TLSSocket files

Commit:
Add license headers into TLSSocket unittests

SeppoTakalo commented:
@0xc0170 fixed!

0xc0170 commented on features/netsocket/DTLSSocketWrapper.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef DTLSSOCKETWRAPPER_H
```
this file misses the license header

0xc0170 reviewed (changes requested):
Build failures look related, there are unrecognized tokens reported in 2 header files

SeppoTakalo commented:
Oh nice... UTF-8.
This means ` ` is not the same as ` ` first one is space, second one is non-breaking space. You can easily enter non-breaking space when you hold the same keys as when you produce `|` (alt-space on Mac)

Both issues fixed now.

```
100% tests passed, 0 tests failed out of 43



Total Test time (real) =   0.57 sec
```

0xc0170 commented:
/morph build

0xc0170 commented:
^^ known failures , still investigating

0xc0170 commented:
Info: This PR has been re-bundled into a new rollup PR (#8838 ).

No further work is needed here, as once that PR is merged, this PR will also be closed and marked as merged.
If any more commits are made in this PR, this PR will remain open and have to go through CI on its own.
